Add support for party management in the sandbox

rautenrieth-da opened the pull request:
This PR adds support for party management in the sandbox:
- Both the in-memory and the SQL backend track a list of known parties, and implicitly add any party mentioned in a transaction.
- New calls were added to the `IndexService` and the `WriteService` for managing parties. These calls are wired to the above mentioned persistence backends, and to a new API service.
- A corresponding client interface was added to the scala API client.
- An integration test was added for checking that a call to allocate a party succeeds.
- An integration test for the sandbox was added for checking that the sandbox implicitly adds all parties mentioned in a transaction

gerolf-da commented on ledger/ledger-api-integration-tests/src/test/itsuite/scala/com/digitalasset/platform/tests/integration/ledger/api/PartyManagementServiceIT.scala:
```diff
@@ -0,0 +1,74 @@
+          result <- client.allocateParty(Some("party"), Some("hint"))
+          finalParties <- client.listKnownParties()
+        } yield {
+          initialParties.length shouldBe 0
```
This assertion will not hold when running the test against a long running ledger (as in DAML-on-X). Maybe the only assertion you can make is that the party is not contained in `initialParties` but IS contained in `finalParties`.

rautenrieth-da replied:
Good point, this test is not very good. And in addition to the above, it also needs to be able to deal with ledgers that reject the call to allocate a party.

mziolekda commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/SandboxPartyManagementService.scala:
```diff
@@ -0,0 +1,81 @@
+
+import akka.stream.Materializer
+import com.daml.ledger.participant.state.index.v2.{
+  PartyManagementService => StatePartyManagementService
```
Hasn't Gabor mentioned that these constituent interfaces of the Index service will ultimately be prefixed by `Index`? If yes, consider using that convention already

rautenrieth-da replied:
Renamed in https://github.com/digital-asset/daml/pull/1452/commits/4691df2c69cc34bf11a712f83de124b567575b3c

mziolekda commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala:
```diff
@@ -193,7 +144,8 @@ class ActiveContractsManager[ACS](initialState: => ACS)(implicit ACS: ACS => Act
                     acc
-                  })
+                  }),
+                  parties = parties.union(nodeParties)
```
Why is there no `AddTransactionState` in this branch?

rautenrieth-da replied:
The above line is part of the `ats.copy` call, which copies the current `AddTransactionState`.

rautenrieth-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala:
```diff
@@ -177,12 +125,15 @@ class ActiveContractsManager[ACS](initialState: => ACS)(implicit ACS: ACS => Act
-                      AddTransactionState(None, errs + DuplicateKey(gk))
+                      AddTransactionState(None, errs + DuplicateKey(gk), parties.union(nodeParties))
                     } else {
                       ats.copy(acc = Some(acc.addContract(absCoid, activeContract, Some(gk))))
```
The set of parties is not updated here. Should be caught by the upcoming test.

meiersi commented:
Thx @rautenrieth-da . Skimmed and LGTM! Go DAML :)

bethaitman commented on docs/source/support/release-notes.rst:
```diff
@@ -57,6 +57,11 @@ Sandbox
+Ledger
+~~~~~~
+
+- Introduced a new API for party management.
```
This sounds like a significant addition, is there any documentation to go with it?

rautenrieth-da replied:
Good point. @bitonic Do we want to add the new API endpoints to the public docs, presumably at https://docs.daml.com/app-dev/services.html?

bitonic replied:
Yes, however not in this PR. I'd keep it on the low for now until things have stabilized.

mziolekda commented on ledger/ledger-api-integration-tests/src/test/itsuite/scala/com/digitalasset/platform/tests/integration/ledger/api/PartyManagementServiceIT.scala:
```diff
@@ -0,0 +1,82 @@
+      }
+    }
+
+    "allocating parties" should {
```
Could we also have tests
- for missing hint 
- for duplicate party names
- for presence and correctness of displayName

rautenrieth-da replied:
Improved the IT to:
- Allocate a party with and without a hint
- Allocate a party with and without a display name
- Check the correctness of displayName
- Concurrently allocate 100 new parties, checking whether they all get unique names
- Deal with implementations that return `UNIMPLEMENTED` or `INVALID_ARGUMENT`

gerolf-da commented on docs/source/support/release-notes.rst:
```diff
@@ -14,6 +14,12 @@ DAML Integration Kit
 
 - Added new CLI flag ``--all-tests`` to the :doc:`Ledger API Test Tool </tools/ledger-api-test-tool/index>` to run all default and optional tests.
 
+Ledger
```
@cbaatz / @cbaatz-da: should we call this `Ledger` or `Sandbox`?

rautenrieth-da replied:
Technically #1312 was for implementing the party management in the `Sandbox`. But the release notes so far do not mention the addition of the new ledger API endpoints (which is the important bit for the end user), so I put it under `Ledger`.

rautenrieth-da replied:
Changed to `Sandbox`, following slack discussions.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPartyManagementService.scala:
```diff
@@ -0,0 +1,80 @@
+import scala.concurrent.{ExecutionContext, Future}
+
+class ApiPartyManagementService private (
+    readBackend: IndexPartyManagementService,
```
This is not necessarily a read backend. In other services we called this like the service, e.g. `transactionService: IndexTransactionService`.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/PostgresLedgerDao.scala:
```diff
@@ -871,6 +898,56 @@ private class PostgresLedgerDao(
+  private val SQL_INSERT_PARTY =
+    SQL("""insert into parties(party, display_name, ledger_offset, explicit)
+        |select {party}, {display_name}, ledger_end, 'true'
+        |from parameters""".stripMargin)
```
Maybe I'm mistaken, but the ledger_end is updated AFTER inserting the transaction and updating the active contracts, which means the party gets the offset of the previous entry (checkpoint, rejection, ...). Or am I missing something?

rautenrieth-da replied:
From the table definition:
```
  -- True iff the party was added explicitly through an API call
  explicit bool not null,
  -- For implicitly added parties: the offset of the transaction that introduced the party
  -- For explicitly added parties: the ledger end at the time when the party was added
  ledger_offset bigint
```

The above SQL statement is only for explicitly inserting parties through the ledger API, which is done outside of the persistence queue used for ledger entries.
The main purpose of this field is to allow a future migration to insert "party added" ledger entries at the correct place. This will not be an easy migration, as inserting new ledger entries in the middle of the stream means transactions will get new offsets.

gerolf-da replied:
oh, right. sorry I missed that. thank you for the explanation.

gerolf-da commented on ledger/sandbox/src/test/suite/scala/com/digitalasset/platform/sandbox/stores/ledger/ImplicitPartyAdditionIT.scala:
```diff
@@ -0,0 +1,187 @@
+            ImmArray.empty,
+            None
+          )
+        )
```
Can you add cases for the other `GenNode` types as well?

rautenrieth-da replied:
Added a test case for `NodeFetch`. `NodeLookupByKey` should not (and does not, in the current implementation) add any new parties.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveContracts.scala:
```diff
@@ -129,7 +147,8 @@ class ActiveContractsManager[ACS](initialState: => ACS)(implicit ACS: ACS => Act
+                  }),
+                  parties = parties.union(nodeParties)
                 )
               case nlkup: N.NodeLookupByKey.WithTxValue[AbsoluteContractId] =>
```
Do we not need to implicitly create the maintainer parties of the key as well?

rautenrieth-da replied:
Good catch.

rautenrieth-da commented on ledger/ledger-api-client/BUILD.bazel:
```diff
@@ -19,6 +19,7 @@ da_scala_library(
         "//3rdparty/jvm/io/grpc:grpc_netty",
         "//3rdparty/jvm/io/netty:netty_handler",
         "//3rdparty/jvm/org/slf4j:slf4j_api",
+        "//daml-lf/data",
```
Why not? `ledger/ledger-api-client` already depends on `ledger/ledger-api-domain`, which in turn depends on `daml-lf/data`. Type definitions in `com.digitalasset.ledger.api.domain` use `Ref.Party` and not `String`.

rautenrieth-da commented on ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/admin/PartyManagementClient.scala:
```diff
@@ -0,0 +1,43 @@
+
+package com.digitalasset.ledger.client.services.admin
+
+import com.digitalasset.daml.lf.data.Ref.Party
```
I see. I defined the `PartyManagementClient` in terms of `com.digitalasset.ledger.api.domain` types, I will change it to use the generated scalapb types.

By the way, if we don't want to use domain types, should we not change all usages `com.digitalasset.ledger.api.domain.LedgerId` to `String`?

gerolf-da replied:
Yes, that's a fair point. Also, this should really be @leo-da's and @S11001001's call.

S11001001 replied:
`Ref.Party` is extremely convenient to use; if it is the intended meaning you should use it, because it is just a `String` where we've checked that only legal Party characters have been used.

I don't consider our use of `String` for Party in any Scala code to be principled, just that the good work that @remyhaemmerle-da has done tightening up the String subtypes in Engine and elsewhere hasn't propagated here yet.

As for domain types in general, @leo-da and I intend to push the Scala ecosystem _away_ from expressing _anything_ in terms of scalapb types, just as the Engine deals with values and transactions in terms of the `daml-lf/transaction` hand-written ADTs instead of the protoc output. In particular, we very much intend to universally remove the scalapb `Identifier`, and the transitive closure of all scalapb types that use it. So take that prejudice as you will.

In this case it seems preferable to use the domain types instead of scalapb as @rautenrieth-da has done. But _notwithstanding the above_ you may consider whether to use `Ref.Party` and whether to also use other domain types separately.

gerolf-da replied:
Thanks for the explanation and outlook, @S11001001 👍

rautenrieth-da commented:
@mziolekda @gerolf-da I have addressed all review comments, except for the ones regarding the use of domain types, see https://github.com/digital-asset/daml/pull/1452#discussion_r291718215
